Load firewall rules on boot on Centos7 (DEVELOP-1756)

slohse opened the pull request:
We've had problems with the firewall rules not being loaded on boot on CentOS 7 systems. This PR fixes that.

I also stumbled across a problem on RedHat/CentOS that leaves me wondering "How did that ever work?": The tasks file includes the distribution specific variables at the beginning. For Debian that doesn't do anything, as that file is empty. But for Red Hat that loads the `iptables_load_path`, a variable that is meant to be overridden via group_vars/host_vars. But if it is included from the `vars/RedHat.yml` file at the run time of the tasks, then it will always be the value from that file.

slohse committed:
load firewall rules on CentOS 7 and above

slohse committed:
On Red Hat it should be yum of course

slohse committed:
Uh. We're at debugging stage six: "How did that ever work?"

slohse committed:
for some reason ansible tries to evaluate the conditional inside the variable here

slohse committed:
oh. hehe. my mistake

slohse committed:
Travis: update Python and Ansible version

slohse committed:
update parameters to new ansible version

mariya commented on defaults/main.yml:
```diff
@@ -9,7 +9,7 @@ iptables_deny_all : yes                 # Deny all except allowed
 iptables_rules_path: "{{ iptables_confdir }}/rules.v4"  # Path to rule file
 
-iptables_load_path: /etc/network/if-up.d/iptables_load # Set empty to prevent loading
+iptables_load_path: "{% if  ansible_os_family == 'Debian' %}/etc/network/if-up.d/iptables_load{% elif ansible_os_family == 'RedHat' %}/etc/sysconfig/iptables-config{% endif %}" # Set empty to prevent loading
```
Wouldn't it be cleaner to put these in vars/RedHat.yml / Debian.yml? And is the "# Set empty to prevent loading" comment still relevant?

slohse replied:
Yeah, the comment is still true, although I don't really see a scenario where you would want that.

I'm not a 100% sure, but I think from the `vars` and `defaults` in roles, only the `main.yml` file is loaded automatically, whereas in `group_vars` and `host_vars` everything in the folders is loaded (the [variable precedence](https://docs.ansible.com/ansible/latest/user_guide/playbooks_variables.html#understanding-variable-precedence) seems to suggest so, using `*` wildcards in one place and `main.yml` in the other).

So you would have to conditionally load it in the role and then you end up with the problem that I fixed here, that it overrides values you may have set in `host_vars` or `group_vars`

mariya replied:
Ah, ok! I *think* I understand. Let's keep this as-is if that's the best option.

mariya commented on tasks/iptables.yml:
```diff
@@ -15,6 +15,12 @@
+  yum:
+    name: firewalld
+    state: absent
+  when: ansible_os_family == 'RedHat'
```
I always find the ansible "when"s a little difficult to follow. It's a lot easier to see all the tasks relevant to one OS in a separate file. 

I would do it like this, though of course you're welcome to suggest an alternative!

main.yml includes one or more of the following based on OS and distro version:
- debian.yml # for all debian distros
- redhat.yml # for all redhat distros
- redhat7.yml # only when redhat >= 7

slohse replied:
You are right of course. I will do that.

slohse replied:
I think I will instead encapsulate the distro-specific tasks into `block`s. They are interleaved with distro-agnostic tasks and it seems weird to `include_tasks` a file with two tasks, jump back into the main file, do two tasks there, then include another file with one task.

I think doing blocks should still make it easier to understand, and as a bonus I found a bug while looking over this.

mariya replied:
Sounds good!

slohse replied:
I'm gonna resolve this now, as I put the distro-specific tasks in blocks and managed to get rid of the CentOS 7 block altogether.

mariya reviewed (changes requested):
Great work! Just a couple small suggestions to make the role easier to read. Has this been tested in staging and/or prod?

slohse committed:
fix logic around autoloading, style improvements

autoloading was broken for CentOS 7 because the load script would not
have been deployed to iptables_load_path. At the same time, if
iptables_load_path was empty (this disabling autoloading), the systemd
service would still have been deployed, trying to load a script that
wasnt there.

Distribution-specific parts wrapped in blocks, task-style changed to
multiline

slohse committed:
uh. iptables-config is just that - a config file or iptables

maybe we shouldn't overwrite that...

slohse committed:
don't use the same file for rules and load

slohse committed:
okay, I think *this* is the place to put it on RH

slohse committed:
And then we don't even need the special case for CentOS 7

slohse commented:
Okay, I think I've got it now. There is something akin to to Debian's `/etc/network/if-up.d/` (i.e. scripts that are executed when a network interface comes up) on Red Hat, but it's a single script file (`/sbin/ifup-local`) rather than a directory where every script within gets executed. And this works on both CentOS 6 and 7, so I was able to remove the special treatment for CentOS 7.

I tested this with `lims-dev-c6` and `lims-dev-c7` on the CentOS side and `megaqc` on the Debian side.

In our `group_vars`/`host_vars` for most if not all CentOS/SL 6 we set the `iptables_load_path` to `/etc/sysconfig/iptables-config`. As the name suggests, this is actually a configuration file for iptables that for some reason gets executed on boot. It is not really meant to contain a script with rules. It doesn't appear to do any harm that we kind of miss-use it that way, but it's not a very neat solution. Long story short: when we transition systems to CentOS 7, we should remember to remove that variable from the `group_vars`/`host vars` and let the default of this role put it to `/sbin/ifup-local`

slohse committed:
the systemd service file isn't needed anymore

mariya commented:
Let's create a JIRA ticket to clean up the variables. Otherwise looks good!

slohse commented:
Ticket: https://snpseq.atlassian.net/browse/DEVELOP-1809
